Binary instrumentation for dynamic analysis

JohnHamby opened the pull request:
This change set introduces rudimentary instrumentation of C# binaries for dynamic analysis, under control of an EmitOption. Some of the compiler edits are surely wrongheaded--please offer up more appropriate solutions.

@amcasey @KevinH-MS @tannergooding @genlu @tmat @ManishJayaswal @jcouv @VSadov @AlekseyTs @jaredpar @cston and any other interested parties are invited.

The instrumentation strategy (modulo some refinements in method prologues and to the helper class API) is explained in:

https://microsoft.sharepoint.com/teams/managedlanguages/_layouts/OneNote.aspx?id=%2Fteams%2Fmanagedlanguages%2Ffiles%2FTeam%20Notebook%2FRoslyn%20Team%20Notebook&wd=target%28Interactive%20Execution%2FTest%20Impact.one%7CAE81BE1C-65C3-4BB5-BA69-83425EE8D8EE%2FInstrumented%20Binaries%7CA959106B-92E9-458B-828C-82643841A659%2F%29

JohnHamby commented:
This is to go into features/test-impact, not to master or future.

JohnHamby commented on build/Targets/Settings.targets:
```diff
@@ -16,9 +16,6 @@
     <BaseNuGetRuntimeIdentifier Condition="'$(OS)' == 'Windows_NT'">win7</BaseNuGetRuntimeIdentifier>
-
-    <!-- Turn on hard-linking on Windows to reduce copy time and disk space -->
-    <CreateHardLinksForCopyLocalIfPossible Condition="'$(OS)' == 'Windows_NT'">true</CreateHardLinksForCopyLocalIfPossible>
```
This is a fix that has already been made in master.

amcasey replied:
Maybe we should just merge with future?  It looks like you had to fix some version numbers too.

amcasey replied:
Ignore my comment about version numbers - I passed the wrong arguments to diff.

JohnHamby commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -435,6 +435,12 @@
     <Field Name="GetTypeFromHandle" Type="MethodSymbol" Null="allow"/>
   </Node>
 
+  <Node Name="BoundMethodToken" Base="BoundExpression">
```
This is a bound node that represents the metadata token value for a method. The strategy for finding span descriptors for particular methods at instrumentation time depends on tracking method tokens.

amcasey replied:
Consider adding a similar comment to the node declaration.

amcasey replied:
This looks very similar to `BoundMethodToken` below.  We're not sharing the representation because we need to be able to distinguish this use case?  Would it be possible to use the presence or absence of `GetMethodFromHandle` as the distinguishing factor?

JohnHamby replied:
1) Yes. 2) Maybe, if lowering could be prevented from filling in GetMethodFromHandle.

JohnHamby commented on src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:
```diff
@@ -2645,6 +2652,12 @@ private void EmitSizeOfExpression(BoundSizeOfOperator boundSizeOfOperator)
 
+        private void EmitMethodTokenExpression(BoundMethodToken node)
+        {
+            _builder.EmitOpCode(ILOpCode.LoadMethodTokenAsInteger_i4);
```
LoadMethodTokenAsInteger_i4 is a bogus/pseudo opcode that is rewritten during metadata writing to become a Ldc_i4 instruction with the real metadata token of the method as its value.

amcasey replied:
Method tokens are 4 bytes on every platform, correct?

JohnHamby replied:
I confess to having assumed that. Can someone provide a definitive answer?

tmat replied:
Yes. BTW, I don't like the idea of hacking IL instructions. I'll stop by to discuss options.

JohnHamby replied:
I didn't/don't like the idea much either, but couldn't think of a workable alternative. I'd love to learn a better way.

JohnHamby replied:
@tmat has suggested using ldtoken with the high bit set in the fake token value, and rewriting the instruction to be a ldc during metadata emission when the high bit is set. I'm not sure which approach is easier to understand. Opinons?

amcasey replied:
I think it would be easier to debug if we used `-1 * pseudotoken` but, otherwise, that sounds reasonable.

JohnHamby replied:
I've implemented the high bit trick, using bit twiddling rather than negation, so that we can abuse some of the other bits in the high byte later.

JohnHamby commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+        {
+            // ToDo: If the type containing the method is generic, synthesize a helper type and put the payload field there.
+            // If the payload field is part of a generic type, there will be a new instance of the field per instantiation of the generic,
+            // and so the payload field must be a member of another type.
```
Can someone point me at a good way to synthesize a helper class to contain payload fields?

amcasey replied:
Is the helper class specific to the code being compiled?  If not, can we just put it in the shared "runtime" assembly?  If so, most of our code synthesis is in the `LocalRewriter`.

JohnHamby replied:
Yes, the members of the class vary per compilation.

JohnHamby commented on src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs:
```diff
@@ -942,6 +952,16 @@ public BoundExpression Typeof(TypeSymbol type)
+                Syntax,
+                Type(type),
+                // For some reason the rewriting logic asserts that the GetTypeFromHandle property is null.
+                null,
```
It's probably cleaner to remove the assertion in the rewriting logic (and eliminate this method), but I didn't want to do that without the approval of someone in the compiler team.

amcasey replied:
Guess: We use the same representation before and after lowering and the presence or absence of that field is used to distinguish the two cases.

JohnHamby commented on src/Compilers/Test/Utilities/CSharp.Desktop/CSharpTestBase.cs:
```diff
@@ -74,6 +75,7 @@ private Action<IModuleSymbol> Translate(Action<ModuleSymbol> action)
             string expectedOutput = null,
             CompilationOptions options = null,
             ParseOptions parseOptions = null,
+            EmitOptions emitOptions = null,
```
EmitOptions are added to a longish chain of method calls. Is there a better way to turn on an EmitOption for tests that compile and run instrumented binaries?

jaredpar commented:
CC @dotnet/roslyn-compiler

amcasey commented on src/Compilers/Core/Portable/System/Reflection/Metadata/ILOpCode.cs:
```diff
@@ -204,6 +204,9 @@ enum ILOpCode : ushort
         Conv_u = 0xe0,
+        // LoadMethodTokenAsInteger_i4 is a pseudo-opcode used to trick the metadata writer into putting a method token value into a ldc_i4.
+        // Its value must fit in one byte (because it will be overwritten with Ldc_i4) and must not collide with a real opcode.
+        LoadMethodTokenAsInteger_i4 = 0xef,
```
Out of curiosity, why `0xef`?

JohnHamby replied:
0xef is the greatest allowed value, and so is the furthest from expected future conflicts.

VSadov commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:
```diff
@@ -525,7 +525,7 @@ private static ImmutableArray<RefKind> GetRefKindsOrNull(ArrayBuilder<RefKind> r
             {
                 BoundExpression argument = rewrittenArguments[a];
-                int p = (!argsToParamsOpt.IsDefault) ? argsToParamsOpt[a] : a;
+                int p = (!argsToParamsOpt.IsDefaultOrEmpty) ? argsToParamsOpt[a] : a;
```
Why did this change? Default mapping was used to indicate that we do not have any remapping (common case). Otherwise remapping must match the length of parameters and it gets more complicated.

amcasey replied:
+1, what changed?

JohnHamby replied:
Vlad's correction to the new Call method below obviates this.

VSadov commented on src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs:
```diff
@@ -612,6 +612,16 @@ public BoundCall Call(BoundExpression receiver, MethodSymbol method, ImmutableAr
+            return new BoundCall(
+                Syntax, receiver, method, args,
+                ImmutableArray<String>.Empty, argumentRefKinds, false, false, false,
+                ImmutableArray<int>.Empty, LookupResultKind.Viable, method.ReturnType)
```
I think this method and the one above should pass `default(ImmutableArray<int>)` as parameter remapping

JohnHamby replied:
Thanks!

In copying the form of the method above, I copied the (apparently incorrect) value of the argsToParams mapping. This lead me to change the local rewriter to treat empty arrays as equivalent to arrays with default values, figuring that I was fixing a bug.

amcasey commented on src/Compilers/CSharp/Portable/BoundTree/Expression.cs:
```diff
@@ -1809,6 +1809,21 @@ public override void Accept(OperationVisitor visitor)
+    {
+        protected override OperationKind ExpressionKind => OperationKind.None;
+
+        public override void Accept(OperationVisitor visitor)
```
This should be unreachable, correct?  Since we don't expose lowered trees through IOperation?

JohnHamby replied:
It's unreachable today. This node, along with others, will need a real implementation when we decide to expose IOperation over lowered trees.

amcasey replied:
Until then, should it throw `NotSupportedException` or `ExceptionUtilities.Unreachable`?  At the very least, I would expect it to assert.

JohnHamby replied:
All of the other unexposed node types act this way. A systematic change to all of them might be worthwhile, but is outside the scope of this change set.

ManishJayaswal replied:
Can we at-least file an issue to track this then?

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed class Instrumentation
```
`static`?

JohnHamby replied:
Apparently so.

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+
+        internal static BoundBlock InjectInstrumentation(MethodSymbol method, BoundBlock methodBody, int methodOrdinal, TypeCompilationState compilationState, CSharpCompilation compilation, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider, out ImmutableArray<SourceSpan> dynamicAnalysisSpans)
+        {
+            if (methodBody != null && (object)method != GetCreatePayload(compilation) && (object)method != GetFlushPayload(compilation))
```
Why is it safe to use identity for the method checks?  If `method` is a definition, can you please add an assert (and, ideally, an explanation) to that effect?

JohnHamby replied:
The checks are not necessary in the real world, in which the helper class comes from a referenced assembly. They are needed only when the helper class definition is part of the current compilation. (In particular, if GetCreatePayload is instrumented it goes into an infinite recursion.) I'll write something in the source text...

amcasey replied:
Also, consider inverting this if.

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+            {
+                MethodSymbol createPayload = GetCreatePayload(compilation);
+                MethodSymbol flushPayload = GetFlushPayload(compilation);
+                if (createPayload != null && flushPayload != null)
```
Please cast symbols to `object` for null checks.

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+
+                    // ToDo: The containing module's mvid should be computed statically and stored in a static field rather than being
+                    // recomputed in each method prologue.
+                    BoundExpression mvid = factory.Property(factory.Property(factory.TypeofBeforeRewriting(method.ContainingType), "Module"), "ModuleVersionId");
```
Is this property accessed once for each method?  Would it make more sense to access it once for the whole assembly?

amcasey replied:
Ah, just noticed your TODO.

amcasey replied:
Consider filing a catch-all bug like "Improve instrumentation perf" and including the URL in the TODO.

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+            // and so the payload field must be a member of another type.
+            NamedTypeSymbol containingType = method.ContainingType;
+
+            SynthesizedFieldSymbol payloadField = new SynthesizedFieldSymbol(containingType, payloadType, method.Name + "*instrumentation*" + methodOrdinal.ToString(), isStatic: true);
```
Please add a new method to the `GeneratedNames` helper class.  That's the only way we can keep track of all the ways we produce mangled names.  Also, is "*" really allowed in method names?

JohnHamby replied:
Ah, I didn't know about GeneratedNames. As for the '*' characters, the generated binaries pass peverify.

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+        {
+            if (_createPayload == null)
+            {
+                NamedTypeSymbol instrumentationType = compilation.GetTypeByMetadataName("Microsoft.CodeAnalysis.Runtime.Instrumentation");
```
Normally, I would expect this to go in `WellKnownTypes`.

JohnHamby replied:
Will do.

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+                NamedTypeSymbol instrumentationType = compilation.GetTypeByMetadataName("Microsoft.CodeAnalysis.Runtime.Instrumentation");
+                if (instrumentationType != null)
+                {
+                    ImmutableArray<Symbol> createPayloads = instrumentationType.GetMembers("CreatePayload");
```
Ditto for `WellKnownMembers`.

JohnHamby replied:
Will do.

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+                if (instrumentationType != null)
+                {
+                    ImmutableArray<Symbol> createPayloads = instrumentationType.GetMembers("CreatePayload");
+                    if (createPayloads.Length == 1)
```
Produce some sort of diagnostic if we can't find the member we need?

JohnHamby replied:
OK.

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+                    if (createPayloads.Length == 1)
+                    {
+                        MethodSymbol createPayload = createPayloads[0] as MethodSymbol;
+                        // ToDo: Add checks for parameter types.
```
Please include a bug number in each TODO.  In this case, the blanked bug for implementing instrumentation would suffice.

JohnHamby replied:
This evaporates with making the method a WellKnownMember.

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+                    {
+                        MethodSymbol createPayload = createPayloads[0] as MethodSymbol;
+                        // ToDo: Add checks for parameter types.
+                        if (createPayload != null && createPayload.IsStatic && createPayload.ParameterCount == 4 && createPayload.Parameters[0].Name == "mvid" && createPayload.Parameters[1].Name == "methodToken" && createPayload.Parameters[2].Name == "payload" && createPayload.Parameters[3].Name == "payloadLength")
```
Null check without cast.

JohnHamby replied:
Code goes away with making the method a WellKnownMember.

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+        private readonly DiagnosticBag _diagnostics;
+        private readonly DebugDocumentProvider _debugDocumentProvider;
+
+        public InstrumentationInjectionWalker(MethodSymbol method, ArrayBuilder<SourceSpan> spansBuilder, FieldSymbol payload, TypeCompilationState compilationState, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider)
```
Consider making the constructor private and adding a helper that manages the lifetime of the `ArrayBuilder` and returns the desired `ImmutableArray`.

JohnHamby replied:
OK.

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+        }
+    }
+
+    internal sealed class InstrumentationInjectionWalker : BoundTreeRewriterWithStackGuard
```
Is this a walker or a rewriter?

JohnHamby replied:
Both. I'll change the name to include "Rewriter".

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+            _debugDocumentProvider = debugDocumentProvider;
+        }
+
+        public override BoundNode VisitBadStatement(BoundBadStatement node)
```
I feel like I'm probably missing something obvious, but why not override `Visit` rather than each individual `VisitXStatement`?  Wouldn't that make the type more resilient against the addition of new kinds of statements?

JohnHamby replied:
I had mistakenly convinced myself that overriding the individual Visit* methods was a general pattern in the compiler. Overriding Visit simplifies things and is more future proof.

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+            Location statementLocation = statement.Syntax.GetLocation();
+            FileLinePositionSpan lineSpan = statementLocation.GetMappedLineSpan();
+            string path = lineSpan.Path;
+            if (path == "")
```
I vaguely recall that `path.Length == 0` might be preferable.

VSadov commented on src/Compilers/Core/Portable/System/Reflection/Metadata/ILOpCode.cs:
```diff
@@ -204,6 +204,9 @@ enum ILOpCode : ushort
         Leave_s = 0xde,
         Stind_i = 0xdf,
         Conv_u = 0xe0,
+        // LoadMethodTokenAsInteger_i4 is a pseudo-opcode used to trick the metadata writer into putting a method token value into a ldc_i4.
```
This feels a bit fragile. I wonder why not just use RuntimeMethodHandle?

VSadov replied:
- When metadata writer separates from Roslyn and becomes a part of System.Metadata, is this going to be a documented and supported behavior?
- Wheat happens if the IL is subsequently rewritten and token change? 
- Is this supposed to work with .Net native?

amcasey commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,350 @@
+
+            SyntheticBoundNodeFactory statementFactory = new SyntheticBoundNodeFactory(_method, statement.Syntax, _compilationState, _diagnostics);
+            BoundArrayAccess payloadCell = statementFactory.ArrayAccess(statementFactory.Field(null, _payload), statementFactory.Literal(spansIndex));
+            BoundExpressionStatement cellAssignment = statementFactory.Assignment(payloadCell, statementFactory.Literal(true));
```
Would it be worthwhile to have the payload returned from a function so that we have a hook when we generalize it?

JohnHamby replied:
Still thinking about this. The details of instrumentation will change in various ways, and I'll work through it when there is a little more time to ponder things.

amcasey commented on src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs:
```diff
@@ -1293,6 +1293,11 @@ public override BoundNode VisitLiteral(BoundLiteral node)
 
+        public override BoundNode VisitMethodToken(BoundMethodToken node)
+        {
+            return null;
```
Unreachable?

JohnHamby replied:
Nope. I discovered a need to override this method by failing an assertion in a default path.

This method will disappear if we switch to the high-bit trick suggested by @tmat . I'm warming up to that.

AlekseyTs replied:
This probably means that instrumentation is done too early. It should be done after flow analysis.

AlekseyTs replied:
Though, Async rewriter might get us here.

JohnHamby replied:
Instrumentation is done after flow analysis.

VSadov replied:
My thoughts on this:
- Instrumentation must be a separate pass. This way instrumentation would not impact performance in regular scenarios.
- Instrumentation should use unlowered tree since it is isomorphic to the code

AlekseyTs replied:
If performance is a concern, which is premature at this point, there are many different ways to mitigate it (for example, using specialized subclass of LocalRewriter when instrumentation is needed). However, I would be surprised if a check on a boolean field or a null check on a field (if we would pass an optional "statement instrumenter" to the LocalRewriter) would create any noticeable performance impact.

amcasey commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicAnalysisDataReader.cs:
```diff
@@ -174,6 +174,11 @@ private static int ReadReference(ref BlobReader reader, bool smallRefSize)
 
         public ImmutableArray<LinePositionSpan> GetSpans(BlobHandle handle)
         {
+            if (handle.IsNil)
```
When does this happen?

amcasey replied:
When does this happen?

JohnHamby replied:
This happens if a method has no spans. Without this check the method crashes later.

amcasey commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicCodeCoverageTests.cs:
```diff
@@ -0,0 +1,626 @@
+    }
+}
+
+namespace Microsoft.CodeAnalysis.Runtime
```
Can this be extracted into a constant and shared across test cases?

JohnHamby replied:
Yup.

amcasey commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicCodeCoverageTests.cs:
```diff
@@ -0,0 +1,626 @@
+
+            if (System.Threading.Interlocked.CompareExchange(ref payload, new bool[payloadLength], null) == null)
+            {
+                int methodIndex = methodToken & 0xffffff;
```
Does it matter that tokens are 1-indexed?

JohnHamby replied:
No--slot 0 will never get filled in, but this is only test code.

amcasey commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicCodeCoverageTests.cs:
```diff
@@ -0,0 +1,626 @@
+        {
+            if (_mvid != mvid)
+            {
+                _payloads = null;
```
Assuming `mvid` is never `default(Guid)`, couldn't we just initialize `_payloads` here and delete the `if` below?

amcasey replied:
Do we have some reason to assume that there's never a call in-flight when the mvid changes?

JohnHamby replied:
Re in-flight calls: the test has no parallelism, so there are no in-flight calls.

amcasey replied:
Oh, I figured this was a prototype of the code for the runtime assembly.  If everything is sequential in the tests, why do we need to use `CompareExchange`?

JohnHamby replied:
No, this isn't prototype code for the real implementation--it's just test scaffolding.

CompareExchange is not strictly necessary here. An early iteration of the synthesized instrumentation code did a CompareExchange inline. When I moved that to the helper method, I kept the code in the same form because it matched how I was thinking. If it's bothersome we can minimize the code.

amcasey commented:
Done first pass.  I'll take another look once the compiler team has reviewed it.

JohnHamby commented:
I've pushed a new commit that addresses most of the review feedback. I still need to add bug numbers to ToDo comments, and add PROTOTYPE comments in a few places.

Also, if the new tests included in this PR continue to fail in Jenkins, I need to figure out what's different between how Jenkins runs tests and how they run locally.

jcouv commented:
Thanks Andrew for the initial pass. I'm going through the code as we speak.

CC @dotnet/roslyn-compiler for another review. Thanks

jcouv commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,216 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
```
Multiple un-necessary usings. Also sort.

JohnHamby replied:
Unused usings removed. Remaining usings look sorted to me (modulo System coming first).

JohnHamby commented:
I've pushed commits that correct the test failures and include documentation of ToDos.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:
```diff
@@ -2645,6 +2652,12 @@ private void EmitSizeOfExpression(BoundSizeOfOperator boundSizeOfOperator)
+        private void EmitMethodTokenExpression(BoundMethodToken node)
+        {
+            _builder.EmitOpCode(ILOpCode.Ldtoken);
+            EmitSymbolToken(node.Method, node.Syntax, null, encodeAsRawToken: true);
```
Is this supposed to be done only for definitions? Consider asserting that the method is definition then. It would be good to assert the result type of the node, unless there is a better place to do this.

JohnHamby replied:
Yes and yes.

jcouv commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,216 @@
+                    InstrumentationInjectionRewriter.InstrumentMethod(method, methodBody, payloadField, compilationState, diagnostics, debugDocumentProvider, out dynamicAnalysisSpans, out newMethodBody);
+
+                    // Synthesize the initialization of the instrumentation payload array. It should be done either statically or with concurrency-safe code:
+                    //
```
The injected IL is not thread-safe yet. Please mark with PROTOTYPE for follow-up.

JohnHamby replied:
The CreatePayload method is required to deal with concurrency. What concurrency issues are present in the injected code?

jcouv replied:
I see. So we can guarantee both of those constraints?
1. payloadField is atomically null or is ready, but nothing in between
2. Multiple calls to Instrumentation.CreatePayload don't cause harm

JohnHamby replied:
Yes, precisely.

jcouv replied:
Thanks

AlekseyTs commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,212 @@
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal static class Instrumentation
```
Consider adding a doc comment.

JohnHamby replied:
OK. Initially the comment will be rather short.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,212 @@
+{
+    internal static class Instrumentation
+    {
+        internal static BoundBlock InjectInstrumentation(MethodSymbol method, BoundBlock methodBody, int methodOrdinal, TypeCompilationState compilationState, CSharpCompilation compilation, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider, out ImmutableArray<SourceSpan> dynamicAnalysisSpans)
```
Can we get compilation from compilationState?

JohnHamby replied:
Yes.

AlekseyTs replied:
Consider removing the parameter then.

JohnHamby replied:
I removed the parameter before answering the question.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,212 @@
+                {
+                    // Create the symbol for the instrumentation payload.
+                    SyntheticBoundNodeFactory factory = new SyntheticBoundNodeFactory(method, methodBody.Syntax, compilationState, diagnostics);
+                    TypeSymbol boolType = compilation.GetSpecialType(SpecialType.System_Boolean);
```
There is no guarantee the type is present and good, see if factory has a helper that checks that.

JohnHamby replied:
I couldn't find a helper. Other parts of the C# compiler, in particular the local rewriter, use types obtained this way without any validation.

AlekseyTs replied:
factory.SpecialType(...)?

JohnHamby replied:
Lo and behold. Apparently I didn't search very effectively.

I'll use the factory helper. If this is an issue here, then there are a number of places in the local rewriter that should be changed.

AlekseyTs replied:
Sometimes we rely on the fact that "goodness" of a type is checked during binding because feature relies on it elsewhere, which is not the case here.

jcouv commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,216 @@
+                    ImmutableArray<BoundStatement> newStatements = newMethodBody.Statements.Insert(0, payloadIf);
+                    newMethodBody = newMethodBody.Update(newMethodBody.Locals, newMethodBody.LocalFunctions, newStatements);
+
+                    if (IsTestMethod(method))
```
General comment on the instrumentation design: Might this method of flushing payloads be brittle?
- Do we expect there to be a reliable and general way of detecting test methods? 
- Can you ensure the test code will be re-built when needed?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,212 @@
+            return methodBody;
+        }
+
+        private static FieldSymbol GetPayloadField(MethodSymbol method, int methodOrdinal, TypeSymbol payloadType, SyntheticBoundNodeFactory factory, CSharpCompilation compilation)
```
Can we get to the compilation through the factory?

JohnHamby replied:
The compilation parameter turns out to be unreferenced, so I will delete it.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,212 @@
+            // ToDo PROTOTYPE (https://github.com/dotnet/roslyn/issues/9810):
+            // If the type containing the method is generic, synthesize a helper type and put the payload field there.
+            // If the payload field is part of a generic type, there will be a new instance of the field per instantiation of the generic,
+            // and so the payload field must be a member of another type.
```
The other type won't be generic, is this correct?

JohnHamby replied:
Correct.

AlekseyTs replied:
I think all these fields should be placed into PrivateImplementationDetail type where we keep other special things like array initialization blobs, etc. The fact whether instrumentation is done or not shouldn't affect the shape of user types because it might case user code to break. For example, if it reflects over the type and expects to find certain number of fields, etc.

amcasey replied:
I believe that's the plan and this is just an attempt to get something checked in, but @JohnHamby can confirm.

AlekseyTs replied:
That is fine, since we already have a TODO and an issue tracking the work. I am just saying that it probably makes sense to put all payload fields there, regardless whether the type is generic or not.

ManishJayaswal replied:
Why would someone want to reflect over instrumented types? Even if they do, I don't think they should expect to find things as defined in user code because it is by definition, instrumented. I am not against putting these fields in another type but am just curious if there is some scenario that I am missing?

AlekseyTs replied:
This is user type. If I define a type, I am free to reflect over it, instrumentation is an implementation detail of a feature that has nothing to do with intended usage of the type, it is not a language feature.

JohnHamby replied:
@AlekseyTs is correct--adding stuff to types in the program is incorrect. I will make the suggested change in a follow-up PR. That it fixes the generics issue is a nice side effect.

VSadov replied:
We occasionally add static fields to user types, for lambda caches for example.

However we never add instance fields to user types. That can have impact on allocation footprint and in a case of structs could be breaking their layout or making them no longer valid to use in unsafe code.

Basically. synthesized static fields in user types are somewhat ok, but instance fields are not ok.

jcouv commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,212 @@
+                        // finally
+                        // {
+                        //     Instrumentation.FlushPayload();
+                        // }
```
Will `FlushPayload` be able to obtain the information it needs (assembly name, name of the test, etc) absent parameters?

JohnHamby replied:
I'll need to add a parameter or two after I get clear on the requirements with @amcasey .

AlekseyTs commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,212 @@
+        {
+            MethodSymbol helper = compilation.GetWellKnownTypeMember(member) as MethodSymbol;
+
+            if ((object)helper == null)
```
I believe this check is insufficient, and this method duplicates code that already exists. Consider using static helper from the Binder class.

JohnHamby replied:
Will do. I didn't know about the helper. Thanks!

jcouv commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,212 @@
+                        break;
+                    default:
+                        return CollectDynamicAnalysis(base.Visit(node));
+                }
```
I'm wondering whether default in or default out will be more maintainable as the compiler evolves.
@AlekseyTs @gafter, what do you think?

jcouv replied:
Also, for my education, what was the reason to do this `switch` rather than directly do the check that appears below?

```CS
      BoundStatement statement = node as BoundStatement;
           if (statement != null)
```

AlekseyTs commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,212 @@
+        private readonly DiagnosticBag _diagnostics;
+        private readonly DebugDocumentProvider _debugDocumentProvider;
+
+        public static void InstrumentMethod(MethodSymbol method, BoundBlock methodBody, FieldSymbol payloadField, TypeCompilationState compilationState, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider, out ImmutableArray<SourceSpan> dynamicAnalysisSpans, out BoundBlock newMethodBody)
```
Consider returning instrumented body instead.

JohnHamby replied:
There are two returned results of equal importance. In such circumstances, do we generally prefer to return one and have an out parameter for the other?

AlekseyTs replied:
We generally use out parameters only if method already returns something else and there is no good reason to introduce a type to package things together.

JohnHamby replied:
OK.

jcouv commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -974,6 +977,12 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
+                ImmutableArray<SourceSpan> dynamicAnalysisSpans = ImmutableArray<SourceSpan>.Empty;
+                if (_moduleBeingBuiltOpt.EmitOptions.EmitDynamicAnalysisData)
+                {
+                    flowAnalyzedBody = Instrumentation.InjectInstrumentation(methodSymbol, flowAnalyzedBody, methodOrdinal, compilationState, _compilation, _diagnostics, _debugDocumentProvider, out dynamicAnalysisSpans);
```
From my discussion with Neal this morning, I thought we do most of the local lowering in a single scan through the BoundNodes. @gafter, it is ok to have this separate step?

JohnHamby replied:
Instrumentation is not lowering, and should occur before lowering.

AlekseyTs replied:
So far I haven't seen anything that would prevent us from doing instrumentation as part of the lowering pass. The location information should be accurate, after all this is were we add sequence points.

jcouv replied:
From discussion with @AlekseyTs, the recommendation is to move this into `LowerBodyOrInitializer` for now, and consider integrating with `LocalRewriter` later.

AlekseyTs replied:
Even the move to LowerBodyOrInitializer can be done later, I think, as long as we track this work.

JohnHamby replied:
My real concern is that the trees presented to the instrumentation logic appear to be unlowered. @AlekseyTs has explained that that is simple to do. I'll open an issue to track this.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs:
```diff
@@ -608,7 +608,17 @@ public BoundCall Call(BoundExpression receiver, MethodSymbol method, ImmutableAr
+            { WasCompilerGenerated = true };
+        }
+
+        public BoundCall Call(BoundExpression receiver, MethodSymbol method, ImmutableArray<RefKind> argumentRefKinds, ImmutableArray<BoundExpression> args)
```
I couldn't find where this method is referenced from.

JohnHamby replied:
Apparently whatever need I had of it evaporated, because it was unreferenced. I will delete it.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs:
```diff
@@ -1098,6 +1127,15 @@ public BoundExpression Array(TypeSymbol elementType, ImmutableArray<BoundExpress
                 Compilation.CreateArrayTypeSymbol(elementType));
         }
 
+        public BoundExpression Array(TypeSymbol elementType, int elementCount)
```
I couldn't not find references to this new method.

JohnHamby replied:
This was used when the initialization of payload arrays was done inline in method prologues. After that initialization moved to a helper method, this method became unreferenced. I will delete it.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/ILBuilderEmit.cs:
```diff
@@ -43,9 +43,14 @@ internal void EmitToken(string value)
+            // Setting the high bit indicates that the token value is to be interpreted literally rather than as a handle.
+            if (encodeAsRawToken)
+            {
+                token |= 0x80000000;
```
Consider asserting that we will not end up with 0xffffffff

JohnHamby replied:
I'm not sure what a meaningful assertion would be here. The line above, which apparently uses 0xFFFF as an invalid token value, troubles me because 65535 seems like too low a limit on the number of valid tokens. Why isn't the invalid value 0xFFFFFF or 0xFFFFFFFF? @tmat ? @VSadov ?

AlekseyTs commented on src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs:
```diff
@@ -4572,11 +4572,25 @@ private void WriteMethodBodyIL(BlobBuilder builder, IMethodBody methodBody)
+                                if ((pseudoToken & 0x80000000) != 0 && (uint)pseudoToken != 0xffffffff)
+                                {
+                                    writer.Offset = offset - 1;
+                                    writer.WriteByte((byte)ILOpCode.Ldc_i4);
```
Consider asserting what instruction we are overwriting.

JohnHamby replied:
OK.

jcouv commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:
```diff
@@ -0,0 +1,513 @@
+
+    async static Task<string> First(string s)
+    {
+        string result = await Second(s) + ""XXX"";
```
Replace with some other arbitrary string.

JohnHamby replied:
OK.

AlekseyTs commented on docs/features/dynamic-analysis.work.md:
```diff
@@ -0,0 +1,12 @@
+
+- Generate code to compute MVIDs less often than once per invoked method.
+- Synthesize helper types to contain fields for analysis payloads so that methods of generic types can be instrumented.
+- Accurately identify which methods are tests.
```
Should probably add that we need to check handling of constructor initializers and field initializers.

JohnHamby replied:
Yes.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -0,0 +1,212 @@
+            int spansIndex = _spansBuilder.Count;
+            _spansBuilder.Add(new SourceSpan(_debugDocumentProvider.Invoke(path, ""), lineSpan.StartLinePosition.Line, lineSpan.StartLinePosition.Character, lineSpan.EndLinePosition.Line, lineSpan.EndLinePosition.Character));
+
+            // Generate "_payload[pointIndex] = true".
```
We should decide what behavior do we want under debugger. For example, when debugger stops at the statement in source, do we want this assignment happen, or do we want it to happen when user steps to execute the statement. I assume we would want the latter, because user might move instruction pointer to a different statement and resume execution. In this case we would do assignment for a statement that wasn't executed and likely wouldn't do it for the statement that would. So, it feels like we should do something special around sequence points. More reasons to move this code to LocalRewriter and coordinate with creation of sequence points.

JohnHamby replied:
We haven't yet designed interactions between dynamic analysis and debugging, but I agree with @AlekseyTs that injected code should be treated as part of the associated statement.

VSadov replied:
There could be other ways to make sure that instrumentation is treated more like a preamble of a statements. For example instrumentation pass could introduce instrumentation nodes that are wrappers of the statements and move syntactical info from the underlying statement to the instrumenting node (making inner statement appear synthesized) - this way the whole thing would get a sequence point and special handling in lowering would be none or minimal..

JohnHamby commented:
I have pushed a commit that addresses most of the remaining issues, and respectfully request some signoffs.

Thanks everyone for giving this PR such careful attention. I learned some things.

jcouv commented on src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs:
```diff
@@ -6,29 +6,34 @@
                     // Create the symbol for the instrumentation payload.
                     SyntheticBoundNodeFactory factory = new SyntheticBoundNodeFactory(method, methodBody.Syntax, compilationState, diagnostics);
-                    TypeSymbol boolType = compilation.GetSpecialType(SpecialType.System_Boolean);
+                    TypeSymbol boolType = factory.SpecialType(SpecialType.System_Boolean);  compilation.GetSpecialType(SpecialType.System_Boolean);
```
Looks like an edit mistake.

AlekseyTs commented on docs/features/dynamic-analysis.work.md:
```diff
@@ -0,0 +1,14 @@
+- Synthesize helper types to contain fields for analysis payloads so that methods of generic types can be instrumented.
+- Accurately identify which methods are tests.
+- Integrate dynamic analysis instrumentation with lowering.
+- Verify that instrumentation covers constructor initializers and field initializers.
```
We might need to track field initializers separately (probably based on field tokens) because they are (the initializers) might end up getting injected into multiple constructors, i.e. the same code is getting duplicated.

AlekseyTs commented:
LGTM, modulo the edit mistake found by @jcouv.

jcouv commented:
LGTM

VSadov commented:
LGTM

jcouv commented:
One more thing:
I saw an email from last week which captured the new rules for compiler feature work.
-	No TODOS.  Use PROTOTYPE instead

The idea is that we will not allow PROTOTYPE beyond the features branches (we can check when integrating branches). Whereas we won’t be able to check for TODOs because there are pre-existing ones.
Please update your PR before merging.

agocke commented on docs/features/dynamic-analysis.work.md:
```diff
@@ -0,0 +1,14 @@
+
+TODO:
+
+- Generate code to compute MVIDs less often than once per invoked method.
```
Can you change the markup to check boxes instead of bullets so we can keep track of where we started, where we are now, and how far we have to go?

JohnHamby replied:
I'll need to learn how to write markdown, and will need to reconstruct the history of what's been done, so that will be in a follow-on PR.

jcouv replied:
See this [example](https://github.com/gafter/roslyn/blob/3c0b3fe0d3eba565b7884908063c68ffec92c090/docs/features/records.md).
Haven't tried it yet, but I think it's two square brackets with a space or an x in the middle. [ ] or [x]

JohnHamby replied:
Oh. I thought it involved writing some actual markup. Even I can type '[' and ']' characters...
